Update RSE in R learning goals and narrative (#337)
merely-useful/merely-useful.github.io: r-rse/course-design.md

    (unit testing, continuous integration, version control)
    - Is designed to involve the community, such as gaining potential contributors
    (including licensing, code of conduct, contributing guidelines, 
    well described README)
    - Has user-friendly (i.e. targetted to a novice to intermediate user)
    documentation and tutorials on how to use the package (e.g. a nice website)
    - Is adequately structured to make package management and maintanence easier
    (e.g. with Issue Templates, Project Boards, Labels, contributing guidelines)
    
- **Teaching/learning activities**:
    - In class portion: 
        - Some lectures for those material that aren't coding heavily
        (e.g. for lessons to involve community, being a good community member)
        - Live coding (with instructor)
        - Exercises that expand on material covered with instructor that 
        solve a provided problem, with or without scaffolding
    - Out of class: 
        - Group work/project creating an R package based on a set of (given?) problems
        - Formative assessments like coding with immediate feedback on correct/incorrect

>         (via learnr package? Esp with new update to RStudio in version 1.3)

Reviewer: This would be awesome! But, writing correctness tests is time consuming...

        - Reading assignments?

- **Assessment and feedback activities**:
    - Deadlines for milestones to acheive for project, with instructor feedback on progress
    - Out-of-class formative assessments based on maybe learnr tutorials?
    - After completing in-class exercises, go over them as class?

- Key resources to use for this:
  - R packages: usethis, devtools, drake?, pkgdown, testthat, styler, rmarkdown
  - Online books: [R Packages](https://r-pkgs.org/) 
  (we differentiate from this amazing book by structuring everything around 
  this being a course, with exercises, assessments, (group?) projects, 

>   instructions/comments/notes for instructors, etc)

Reviewer: Right, I think of R packages as being a reference book, we are writing for people learning this for the first time.

## Chapter outline

Course chapters and content need to be strongly aligned with learning outcome/goal.
Here is a draft outline to guide development.

- Syllabus
    - Overview of course
    - Why to learn it
    - What makes this course different

> 

Reviewer: I'm a big fan of the "Whole Game" chapter in R Packages: it removes some of the magic/fear around packages by walking through making one from scratch.  I think it's an activity that's doable in less than an hour and gives students a quick win.  Is this something we could do up front?  Or is this what you imagine "Setting up an R package" looks like?

- Folder and file structure (RStudio R Projects)
    - Filepaths, very basics of the shell (in RStudio)
    - Making use of the fs package for filesystem management
    

> - Setting up an R package

Reviewer: Does this start with what a package is, and why you might want to make one?

    - Using devtools, usethis for development
        - e.g. `use_devtools()` and `use_usethis()`.

>     - Developing functions

Reviewer: Can you elaborate? Is this the mechanics of turning a code snippet into a function?  Or is it something more?

    - Making and using datasets

- Function development in a package environment
    - Dependency management
    - Function documentation (with roxygen2), part 1
        - To use `@examples` to help with creating function

- (Idea) Additional features to make use of
    - RStudio addins, snippets, rmarkdown templates, templates

>     - rstudioapi?

Reviewer: You mean as possible components of a package, right?  I think of these as things beyond functions that an app can provide. Having a survey of them in one place would be really nice (I can't think of anywhere else that does).   

Would shiny apps fit here too?

- Version control: Using Git and GitHub as a sole user (part 1)
    - Using Git in RStudio (standard add-commmit-history)
    - Using Git in the Terminal of RStudio (moving in history with checkout,
    creating branches, adding and updating remotes)
    - Setting up GitHub for R package (make use of usethis), pushing and pulling

- Checking correctness of code
    - Unit test (with testthat)
    - Code coverage (with covr)

>     - With devtools

Reviewer: Are our learners already OK with debugging?  E.g. I'm thinking about Jenny Bryan's [Object of closure is not sub-settable keynote](https://rstudio.com/resources/rstudioconf-2020/object-of-type-closure-is-not-subsettable/), where unit tests are the last strategy she covers, do we assume the learners already know the earlier strategies?

- Build management
    - With devtools

>     - Maybe with drake?

Reviewer: My knowledge of drake is minimal...for some reason I've thought about it more as a tool for automating data pipelines.

    - Function documentation (with roxygen2), part 2
    - learnr tutorials via `use_tutorials()`?
    - Running spell checks and styling (with spelling and styler)

- Involving the community:
    - Designing your package to be used by or contributed to from the community
    (refer to Mozilla Science Labs Open Leaders material)
    - Contributing guidelines 
    - Being inclusive (code of conduct)
    - Licensing
    - NEWS file

- Exposing your package to the world with a website:
    - Setting up a website (with pkgdown)
    - Customizing the themes
    - Getting all resources and material organized
    - Defining lifecycle of your package and individual functions (with lifecycle)

- Version control: Workflows around using GitHub (part 2)
    - Make use of R builtin tools like the usethis pull request helpers 
    (and learning how to make use of this yourself)

- Project management of your package development (either sole or in a team)
    - Issue tracking
    - Labeling
    - Project boards
    - Milestones

- Preparing for package release
    - Check builds on other systems (with rhub)
    - Running local CRAN checks
    - Managing your versions, git tagging

> - Making your R package development to a team-based environment

Reviewer: ```suggestion
- R package development in a team-based environment
```

    - Making use of GitHub Organizations/Teams
    - Branch protection, member roles
    - Decision making
    - Governance
    - Issue assignment

>     - Relying on and using code of conduct to build culture and standards

Reviewer: It feels like working on an R package with more than one person doesn't get covered until the end of the book, but in a class setting learners might be working together from very early on in the class.  Can we make this a different "part", and we give some guidance to instructors up front on different times pieces might be covered? E.g. you could imagine covering a classroom Code of Conduct on the first day, and get students to read about why it's a good idea to have one from this section, or when students start on their first group project, you might get them to select and agree to a Code of Conduct for their own group.
